refactor: 1348 upgrade to spring boot 3x

TheGreatRefrigerator opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box
when you have checked you have done them): -->
- [x] 1. I have [**rebased**][rebase] the latest version of the master branch into my feature branch and all conflicts
         have been resolved.
- [x] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the
         [Unreleased] heading.
- [x] 3. I have documented my code using JDocs tags.
- [x] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [x] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [x] 6. I have created API tests for any new functionality exposed to the API.
- [ ] 7. If changes/additions are made to the ors-config.json file, I have added these to the [ors config documentation][config]
         along with a short description of what it is for, and documented this in the Pull Request (below).
- [ ] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [x] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [ ] 10. For new features or changes involving building of graphs, I have tested on a larger dataset
          (at least Germany), and the graphs build without problems (i.e. no out-of-memory errors).
- [ ] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the
          importer etc.), I have generated longer distance routes for the affected profiles with different options
          (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end
          points generated from the current live ORS.
          If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [x] 12. I have written in the Pull Request information about the changes made including their intended usage
          and why the change was needed.
- [x] 13. For changes touching the API documentation, I have tested that the API playground [renders correctly][api].

Fixes # .

### Information about the changes
- Key functionality added: migration to spring-boot 3.1.1
- Reason for change: being up to date
As there were quite some refactorings done, this is not just a chore but refactor PR

### Examples and reasons for differences between live ORS routes, and those generated from this pull request
-

### Required changes to ors config (if applicable)
-

[config]: https://GIScience.github.io/openrouteservice/installation/Configuration.html
[api]: https://gitlab.gistools.geog.uni-heidelberg.de/giscience/openrouteservice-infrastructure/ors-docs-api#test-new-ors-documentation
[rebase]: https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines

TheGreatRefrigerator committed:
chore: migrate to Jakarta EE 9

using new `jakarta` namespace

TheGreatRefrigerator committed:
style(Java17): reformat multiline strings to codeblocks

Co-authored-by: Jakob Schnell <jakob.schnell@heigit.org>

TheGreatRefrigerator committed:
style: use explicit over star import

Co-authored-by: Jakob Schnell <jakob.schnell@heigit.org>

TheGreatRefrigerator committed:
style(Java15): use formatted() instead of String.format()

Co-authored-by: Jakob Schnell <jakob.schnell@heigit.org>

TheGreatRefrigerator committed:
test(Junit5): use default package visibility

Co-authored-by: Jakob Schnell <jakob.schnell@heigit.org>

TheGreatRefrigerator committed:
refactor: remove wrong instanceof checks/casts for RoutingCHGraph

chGraph always is of type RoutingCHGraph, which makes this instance of
check obsolete. QueryGraph is not a subtype of QueryRoutingCHGraph so
this cast is wrong. RoutingCHGraph already has a getEdges() method so
it is just saved in the instance.

- reorder statements for readability

Co-authored-by: Jakob Schnell <jakob.schnell@heigit.org>

TheGreatRefrigerator committed:
style(Java17): use pattern matching with instanceof

Co-authored-by: Jakob Schnell <jakob.schnell@heigit.org>
Co-authored-by: Julian Psotta <22392104+MichaelsJP@users.noreply.github.com>

TheGreatRefrigerator committed:
refactor: cleanup StatusCodeCaptureWrapper.java

after the change from javax to jakarta namespace, the
HttpServletResponseWrapper class was missing the
setStatus(int sc, String sm) declaration resulting in
a compilation error.

As the message, request argument and the statusCode property
were never used anywhere, code for handling them was removed.

- improve docstring
- remove redundant comments

TheGreatRefrigerator committed:
build(deps): use springdoc-openapi-starter-webmvc-ui

the previously uses package springdoc-openapi-ui was not the
correct one to be used with spring-boot 3

the application properties 'springdoc.packages-to-scan' and
'springdoc.pathsToMatch' are now needed for the OpenAPI specs
to be generated properly.

TheGreatRefrigerator committed:
build(deps): use spring-boot-starter-parent v3.1.1

TheGreatRefrigerator committed:
chore: remove unused workaround for springfox-swagger package

as the springfox-swagger package was replaced, this workaround is not
needed anymore.

- optimize imports

TheGreatRefrigerator committed:
refactor: remove GroupedOpenApi builders

as we currently only expose the v2 API we don't need the
grouped OpenAPI specs. Once we want to generate both v2 and v3
API specs in parallel we need to reintroduce this.

TheGreatRefrigerator committed:
style: remove default OpenAPI spec version

the SpecVersion used here should be the OpenAPI V30 which
is already the default for an OpenAPI instance and is therefore removed.

TheGreatRefrigerator committed:
refactor: SwaggerConfig.java to OpenAPIConfig.java

name change due to specs being OpenAPI 3 now instead of swagger 2

koebi committed:
test: re-enable openapi unittest

TheGreatRefrigerator committed:
fix: defaultValue for matrix 'metrics' parameter

TheGreatRefrigerator committed:
test: reintroduce OpenAPI api tests

debugging the exact location of errors was not straight forward,
due to messages containing the wrong location.
Added Hint on how to debug in docstring.
Adjusted size numbers after removal of centrality endpoint.

TheGreatRefrigerator committed:
refactor: name changes from Swagger to OpenAPi

koebi committed:
fix: route request duration format

due to the 'schedule_duration' and 'walking_time' being of
type Duration, the type of the parameters was exposed as objects.
The values passed were however always Strings.

- Adjusted schema type in RouteRequest to String for durations

TheGreatRefrigerator committed:
test: adjust openapi.ResultTest

as the exact number of tags, paths and components will change and for
components is not consistent, we just check if there are some at all.

TheGreatRefrigerator committed:
refactor: provide format info for duration parameters

TheGreatRefrigerator committed:
docs: add CHANGELOG.md entry for spring-boot migration changes

TheGreatRefrigerator commented:
rebased on master

MichaelsJP commented:
@TheGreatRefrigerator Looks like the docker setup is not working as before.
It changes from 000 (I guess no return) to 404. I guess the paths are somehow a bit different now when running the war deployment.

TheGreatRefrigerator commented:
> @TheGreatRefrigerator Looks like the docker setup is not working as before. It changes from 000 (I guess no return) to 404. I guess the paths are somehow a bit different now when running the war deployment.

yes checking rn

koebi committed:
build: upgrade to tomcat 10

springboot 3 requires jakarta servlet 6 which in turn requires tomcat
10:
https://tomcat.apache.org/whichversion.html

TheGreatRefrigerator commented:
@MichaelsJP docker workflow fixed by @koebi 's commit. Ready for review

MichaelsJP commented:
This needs to wait until #1499 is merged. 

Can already be reviewed, nevertheless.

MichaelsJP commented on ors-api/src/main/resources/application.yml:
```diff
@@ -165,3 +165,5 @@ springdoc:
   api-docs:
     path: /v2/api-docs
     version: OPENAPI_3_0
```
Why not 3.1 (for documentation purposes)?

TheGreatRefrigerator replied:
Because with version 3.1 the `type: object` is not generated for parameters. So for now we should stay at Open API 3.0

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/jts/JTS.java:
```diff
@@ -445,10 +445,10 @@ public static Geometry removeCollinearVertices(final Geometry g) {
-        } else if (g instanceof Polygon) {
-            return removeCollinearVertices((Polygon) g);
+        if (g instanceof LineString string) {
+            return removeCollinearVertices(string);
```
```suggestion
        if (g instanceof LineString lineString) {
            return removeCollinearVertices(lineString);
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/jts/JTS.java:
```diff
@@ -287,9 +287,9 @@ private static Geometry smooth(final Geometry geom, final double fit,
             Geometry geom, double fit) {
 
-        if (geom instanceof LinearRing) {
+        if (geom instanceof LinearRing ring) {
```
```suggestion
        if (geom instanceof LinearRing linearRing) {
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/jts/JTS.java:
```diff
@@ -287,9 +287,9 @@ private static Geometry smooth(final Geometry geom, final double fit,
+        if (geom instanceof LinearRing ring) {
             // Treat as a Polygon
-            Polygon poly = factory.createPolygon((LinearRing) geom, null);
+            Polygon poly = factory.createPolygon(ring, null);
```
```suggestion
            Polygon poly = factory.createPolygon(linearRing, null);
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/jts/JTS.java:
```diff
@@ -482,10 +482,10 @@ public static Geometry removeCollinearVertices(final Geometry geometry, int minP
-        } else if (geometry instanceof Polygon) {
-            return removeCollinearVertices((Polygon) geometry);
+        if (geometry instanceof LineString string) {
+            return removeCollinearVertices(string);
```
```suggestion
        if (geometry instanceof LineString lineString) {
            return removeCollinearVertices(lineString);
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/matrix/algorithms/MatrixAlgorithmFactory.java:
```diff
@@ -27,8 +27,8 @@ private MatrixAlgorithmFactory() {
-        if (gh instanceof ORSGraphHopper) {
-            if (req.getSearchParameters().getDynamicSpeeds() && ((ORSGraphHopper) gh).isCoreEnabled()) {
+        if (gh instanceof ORSGraphHopper hopper) {
+            if (req.getSearchParameters().getDynamicSpeeds() && hopper.isCoreEnabled()) {
```
```suggestion
        if (gh instanceof ORSGraphHopper orsGraphHopper) {
            if (req.getSearchParameters().getDynamicSpeeds() && orsGraphHopper.isCoreEnabled()) {
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/core/PrepareCore.java:
```diff
@@ -163,8 +163,8 @@ protected boolean doNotContract(int node) {
-        if (neighbors instanceof IntArrayList)
-            ((IntArrayList) neighbors).removeAll(isCoreNode);
+        if (neighbors instanceof IntArrayList list)
+            list.removeAll(isCoreNode);
```
```suggestion
        if (neighbors instanceof IntArrayList intArrayList)
            intArrayList.removeAll(isCoreNode);
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/LMEdgeFilterSequence.java:
```diff
@@ -61,11 +61,11 @@ private boolean isAvoidCountry(String countries){
 		//Check if the avoidBordersFilter has the same countries or a subset
 		for (EdgeFilter edgeFilter: filters) {
-			if (edgeFilter instanceof AvoidBordersCoreEdgeFilter){
+			if (edgeFilter instanceof AvoidBordersCoreEdgeFilter filter){
```
```suggestion
			if (edgeFilter instanceof AvoidBordersCoreEdgeFilter coreEdgeFilter){
```

MichaelsJP commented on ors-engine/src/main/java/org/heigit/ors/routing/graphhopper/extensions/edgefilters/core/LMEdgeFilterSequence.java:
```diff
@@ -61,11 +61,11 @@ private boolean isAvoidCountry(String countries){
 				if(queryCountries.isEmpty())
 					return false;
-				return queryCountries.containsAll(Arrays.stream(((AvoidBordersCoreEdgeFilter) edgeFilter).getAvoidCountries()).boxed().collect(Collectors.toList()));
+				return queryCountries.containsAll(Arrays.stream(filter.getAvoidCountries()).boxed().collect(Collectors.toList()));
```
```suggestion
				return queryCountries.containsAll(Arrays.stream(coreEdgeFilter.getAvoidCountries()).boxed().collect(Collectors.toList()));
```

MichaelsJP commented:
Just a few minor Style things. Generally, it can become confusing when naming objects close to language level structurs such as `list`, `filter`, `string` etc.

TheGreatRefrigerator commented:
changed some more of the pattern variables, see
https://github.com/GIScience/openrouteservice/compare/d38be18950e9f434f70cb7a4143ec1ffd16cd66d..61523f797ba36b216dc104f8f1e8a4b23816bd0d
